crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
     /// A special ID that is used for an "always true" / "always visible" constraint.
-    /// When we create a new [`VisibilityConstraints`] object, this constraint is always
-    /// present at index 0.
     pub(crate) const ALWAYS_TRUE: ScopedVisibilityConstraintId =
-        ScopedVisibilityConstraintId::from_u32(0);
+        ScopedVisibilityConstraintId(0xffff_ffff);
+
+    /// A special ID that is used for an ambiguous constraint.
+    pub(crate) const AMBIGUOUS: ScopedVisibilityConstraintId =
+        ScopedVisibilityConstraintId(0xffff_fffe);
 
     /// A special ID that is used for an "always false" / "never visible" constraint.
-    /// When we create a new [`VisibilityConstraints`] object, this constraint is always
-    /// present at index 1.
     pub(crate) const ALWAYS_FALSE: ScopedVisibilityConstraintId =
-        ScopedVisibilityConstraintId::from_u32(1);
+        ScopedVisibilityConstraintId(0xffff_fffd);
 
-    /// A special ID that is used for an ambiguous constraint.
-    /// When we create a new [`VisibilityConstraints`] object, this constraint is always
-    /// present at index 2.
-    pub(crate) const AMBIGUOUS: ScopedVisibilityConstraintId =
-        ScopedVisibilityConstraintId::from_u32(2);
+    fn is_terminal(self) -> bool {
+        self.0 >= SMALLEST_TERMINAL
+    }
 }
 
+// Rebind some constants locally so that we don't need as many qualifiers below.
+const ALWAYS_TRUE: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::ALWAYS_TRUE;
+const AMBIGUOUS: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::AMBIGUOUS;
+const ALWAYS_FALSE: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::ALWAYS_FALSE;
+const SMALLEST_TERMINAL: u32 = ALWAYS_FALSE.0;
+
+/// A collection of visibility constraints. This is currently stored in `UseDefMap`, which means we
+/// maintain a separate set of visibility constraints for each scope in file.
 #[derive(Debug, PartialEq, Eq)]
 pub(crate) struct VisibilityConstraints<'db> {
-    constraints: IndexVec<ScopedVisibilityConstraintId, VisibilityConstraint<'db>>,
+    constraints: Vec<Constraint<'db>>,
+    interiors: Vec<InteriorNode>,
 }
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, Default, PartialEq, Eq)]
 pub(crate) struct VisibilityConstraintsBuilder<'db> {
-    constraints: IndexVec<ScopedVisibilityConstraintId, VisibilityConstraint<'db>>,
-}
-
-impl Default for VisibilityConstraintsBuilder<'_> {
-    fn default() -> Self {
-        Self {
-            constraints: IndexVec::from_iter([
-                VisibilityConstraint(VisibilityConstraintInner::AlwaysTrue),
-                VisibilityConstraint(VisibilityConstraintInner::AlwaysFalse),
-                VisibilityConstraint(VisibilityConstraintInner::Ambiguous),
-            ]),
-        }
-    }
+    constraints: Vec<Constraint<'db>>,
+    interiors: Vec<InteriorNode>,
+    constraint_cache: FxHashMap<Constraint<'db>, u32>,
+    interior_cache: FxHashMap<InteriorNode, u32>,
```

Comment:
if we end up not using `IndexVec`, I still recommend to create a newtype wrapper around `u32` to signify what the `u32` here means (it's an index into the constraints/interiors vec?)

Reply:
We're back to using `IndexVec`